internal/commands/request/request_internal.go

```diff
@@ -0,0 +1,294 @@
+package request_internal
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingidentity/pingctl/internal/configuration/options"
+	"github.com/pingidentity/pingctl/internal/customtypes"
+	"github.com/pingidentity/pingctl/internal/output"
+	"github.com/pingidentity/pingctl/internal/profiles"
+)
+
+type PingoneAuthResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+func RunInternalRequest(uri string) (err error) {
+	service, err := profiles.GetOptionValue(options.RequestServiceOption)
+	if err != nil {
+		return fmt.Errorf("failed to send custom request: %v", err)
+	}
+
+	if service == "" {
+		return fmt.Errorf("failed to send custom request: service is required")
+	}
+
+	switch service {
+	case customtypes.ENUM_REQUEST_SERVICE_PINGONE:
+		err = runInternalPingOneRequest(uri)
+		if err != nil {
+			return fmt.Errorf("failed to send custom request: %v", err)
+		}
+	default:
+		return fmt.Errorf("failed to send custom request: unrecognized service '%s'", service)
+	}
+
+	return nil
+}
+
+func runInternalPingOneRequest(uri string) (err error) {
+	accessToken, err := pingoneAccessToken()
+	if err != nil {
+		return err
+	}
+
+	topLevelDomain, err := getTopLevelDomain()
+	if err != nil {
+		return err
+	}
+
+	apiURL := fmt.Sprintf("https://api.pingone.%s/v1/%s", topLevelDomain, uri)
+
+	httpMethod, err := profiles.GetOptionValue(options.RequestHTTPMethodOption)
+	if err != nil {
+		return err
+	}
+
+	if httpMethod == "" {
+		return fmt.Errorf("http method is required")
+	}
+
+	data, err := getData()
+	if err != nil {
+		return err
+	}
+
+	payload := strings.NewReader(data)
+
+	client := &http.Client{}
+	req, err := http.NewRequest(httpMethod, apiURL, payload)
```

Comment:
It might be useful to have a way to print request bodies/headers/etc. for debugging purposes. Would just need to be sure to sanitize the output for any sensitive values.

Reply:
Resolved internally.